src/intrinsics/ecma262/Object.js

```diff
@@ -41,6 +41,92 @@ import { Create, Havoc, Properties as Props, To } from "../../singletons.js";
+  if (!(value instanceof AbstractObjectValue)) {
+    return false;
+  }
+  let realm = value.$Realm;
+  let temporalBuildNodeEntry = realm.getTemporalBuildNodeEntryArgsFromDerivedValue(value);
+  if (temporalBuildNodeEntry === undefined) {
+    return false;
+  }
+  let args = temporalBuildNodeEntry.args;
+  // If we are Object.assigning 3 or more args
+  if (args.length < 3) {
+    return false;
+  }
+  // Then scan through the args after the "to" of this Object.assign, to see if any
+  // other sources are the "to" of a previous Object.assign call
+  for (let i = 2; i < args.length; i++) {
+    let possibleOtherObjectAssignTo = args[i];
+    // Ensure that the "to" value can be omitted
+    if (!callbacks.canOmit(possibleOtherObjectAssignTo)) {
+      continue;
+    }
+    // Check if the "to" was definitely an Object.assign, it should
+    // be a snapshot AbstractObjectValue
+    if (
+      possibleOtherObjectAssignTo instanceof AbstractObjectValue &&
+      // We don't handle conditions for now, this can be done at a later stage
+      possibleOtherObjectAssignTo.kind !== "conditional"
+    ) {
+      let otherTemporalBuildNodeEntry = realm.getTemporalBuildNodeEntryArgsFromDerivedValue(
+        possibleOtherObjectAssignTo
+      );
+      if (otherTemporalBuildNodeEntry === undefined) {
+        continue;
+      }
+      let otherArgs = otherTemporalBuildNodeEntry.args;
+      let objectAssingFunc = realm.intrinsics.Object.$Get("assign", realm.intrinsics.Object);
+      // Object.assign has at least 2 args and the first is that of Object.assign native function
+      if (otherArgs.length < 3 || otherArgs[0] !== objectAssingFunc) {
+        continue;
+      }
+      let to = args[1];
+      // If we cannot omit the "to" value that means it's being used, so we shall not try to
+      // optimize this Object.assign.
+      if (!callbacks.canOmit(to)) {
+        let newArgs = [objectAssingFunc, to];
+        for (let x = 2; x < otherArgs.length; x++) {
+          newArgs.push(otherArgs[x]);
+        }
+        for (let x = 2; x < args.length; x++) {
+          let arg = args[x];
+          // We don't want to add the "to" that we're merging with!
+          if (arg !== possibleOtherObjectAssignTo) {
+            newArgs.push(arg);
+          }
+        }
+        // We now mutate the args in place with the above new args, clearing out the old args.
+        // The end result should be a merged Object.assign and one less temporal entry.
+        // The previous Object.assign temporal's "to" is no longer being used anywhere and should
+        // now dead-code away nicely.
+        args.length = 0;
+        args.push(...newArgs);
+        // Lastly, because we want to possibly merge many Object.assigns, not only two, we do this
+        // process again from the start. Given we've mutated in place the args, this will work.
+        attemptToMergeEquivalentObjectAssigns(callbacks, value);
```

Comment:
So an outermost call to `attemptToMergeEquivalentObjectAssigns` might cause state mutations in some distant generator entries. I really don't like adding this kind of complications to the visitor.

Background: Currently, the visitor is doing its fixpoint computation in a very naive way, re-running all possible contributors to value declarations, being ignorant of actual dependencies. This is already a huge performance issue for the React Compiler in some cases. Eventually, I want this to be dependency driven. But for that to work, we cannot peek into remote dependencies or even rewrite them behind the scenes.

Reply:
We don't mutate other entries, we only mutate the existing entry, recursively merging in the other entries.